feat(price-feeds): add ExpressionPriceFeed

mrice32 opened the pull request:
**Motivation**

To support more arbitrary types of price feeds, it is becoming clear that it would be useful to be able to define arbitrary expressions that combine existing price feeds. This is a proposal for a price feed that takes expressions of the form:

```
"BTCUSD * 100 / ETHUSD"
```

Note: the library can actually handle _far more_ complex expressions, even multiline ones. But, for now, we should encourage users to keep them fairly simple.

**Summary**

This is a _proposal_ for a type of expression-parsing price feed to add more generality to the price feed creation process. This would move much of the simple math to the config and leave only the data-acquisition to the price feeds themselves.

I haven't added the configuration and construction portions of this price feed, as I want to ensure the team agrees with this rough direction. However, it's important to note that I think this price feed will be constructed by searching for known strings in the expression and including price feeds associated with each of those known strings. The user would basically be able to use existing default price feed identifiers _and_ they would be able to include a custom mapping in their config for additional price feed names that they'd like to map.

Ultimately, this expression parser could make the medianizer price feed unnecessary, as the library used would allow us to inject a custom medianizer function.

This will also facilitate the badger price feed.

Other notes:
- The `mathjs` library used for expression parsing is quite powerful and large. It has some very convenient scoping mechanisms that allow us to replace "symbols" with values. This sort of functionality could potentially be replaced with a call to `eval` but that's generally considered bad practice due to security and performance concerns. However, it would shrink the size of the dependencies considerably.
- To make _all_ of our default price identifiers work with this sort of strategy, we'll need to strip them of `/` characters (which could be confused with a division operator by the parser) when ingesting them here. That's not a huge concern, but it is a slight added complexity.
- The decimal strategy in this class is quite different from others. In this case, I use a high-precision BigNumber class that comes with the mathjs library to convert all values to decimal before being processed by the library. This allows the library to use its default arithmetic operators, which is far far simpler than trying to make it operate on fixed point values. To ensure this BigNumber can handle our high-precision fixed point numbers, I customized the config to give them 100 decimals of precision, which should be plenty for our use. After the operations, it rounds off the result to drop any precision that's larger than our fixed point numbers allow.

This is how you configure an expression price feed:
```js
{
  type: "expression",
  expression: "ETHBTC + MYCUSTOMFEED",
  customFeeds: {
    MYCUSTOMFEED: {
      type: "cryptowatch",
      exchange: "coinbase-pro",
      pair: "ethusd"
    }
  }
}
```

In this example, the parser would find that the two "symbols" in the expression are `ETHBTC` (notice that we have stripped the `/` character) and `MYCUSTOMFEED`. Preferring customFeeds in the resolution, we would look through customFeeds _and_ the default feeds in the repo searching for each of these symbols. If any resolution fails, the construction fails. In this case, however, it would construct both feeds and use them to evaluate the expression. The nice thing about the early resolution is that any symbols that aren't in the expression are pruned, so there should never be any unnecessary feeds being constructed and/or evaluated.

Note: for feed names that contain special characters that are allowed in identifiers: `/`, ` `, `[`, `]`, they will need to be escaped, like `"\\/"` in the configuration.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [x]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested


**Issue(s)**

N/A

nicholaspai commented on packages/financial-templates-lib/src/price-feed/ExpressionPriceFeed.js:
```diff
@@ -0,0 +1,117 @@
+class ExpressionPriceFeed extends PriceFeedInterface {
+  constructor(priceFeedMap, expression) {
+    super();
+    this.expressionCode = math.parse(expression).compile();
```
Will this throw if `expression` is invalid? I think thats a good outcome but just want to make sure

mrice32 replied:
It depends on _how_ it is invalid. It will throw if the expression is unparsable, but it will _not_ throw if it's parsable by the engine, but doesn't resolve to a number or if one of the symbols isn't resolved by the scope.

mrice32 replied:
Just to follow-up here. It's difficult to guarantee valid-ness of the expression without passing in values for all of the symbols, but we also can't assume which symbol values would be valid in the expression. For instance, passing in 1 for all values would cause the following to throw:
```
(A + B) / (A - B)
```

mrice32 replied:
Technically, we could check that all symbols are defined in the map, but I don't think we could check for all types of expression errors.

nicholaspai commented on packages/financial-templates-lib/src/price-feed/ExpressionPriceFeed.js:
```diff
@@ -0,0 +1,117 @@
+      throw errors;
+    }
+
+    return this._convertToFixed(this.expressionCode.evaluate(historicalPrices));
```
Does this imply that the `priceFeedMap` has to list pricefeeds in the exact order that they are used in the `expressionCode`?

mrice32 replied:
This is a great question. I will add comments as you suggest and we should probably have more checks around detecting if the output is valid. Here's how this roughly works:
```js
let scope = {
  a: 5,
  b: 6
};
let output = math.parse("a + b").compile().evaluate(scope);
```

In this case output is 11. You can use a and b however you like in the expression and they can be any order in the scope. The scope just defines how it resolves each symbol.

However, I'm not sure what happens in this case:
```js
let scope = {
  a: 5
};
let output = math.parse("a + b").compile().evaluate(scope);
```

My guess is that it would return something like "b + 5", but I need to do more testing to make sure we error gracefully in that case and maybe catch it early.

mrice32 replied:
It's actually nicer @nicholaspai. It throws if any of the symbols are not resolved at runtime.

nicholaspai reviewed (approved):
I really like the overall direction of this PR and 100% agree it will be useful. I'd like to see the following implemented before we merge (which I'm sure you're working on)
- Better comments on new pricefeed class and functions
- Comment above constructor explaining what types of expressions are valid
- Comment explaining how the pricefeed map should be constructed 

One cool thing we can do with this is replace the `inverse` boolean flag from the `Medianizer` and just use this pricefeed instead and define the expression as `1 / MedianizedPriceFeed`

mrice32 commented:
> * Comment explaining how the pricefeed map should be constructed

I think this PR should add to CreatePriceFeed.js to allow users to create/configure this price feed through the config. Will do this and re-request your review.

chrismaree commented on packages/financial-templates-lib/src/price-feed/ExpressionPriceFeed.js:
```diff
@@ -0,0 +1,117 @@
+
+  // Takes a BN fixed point number and converts it to a math.bignumber decimal number that the math library can handle.
+  _convertToDecimal(price) {
+    const decimals = math.bignumber(this.getPriceFeedDecimals());
```
so this limits this feed to only outputting at the decimals of the underlying price feed. It might be worth while to set this as a parameter. for example, there could be a feed that is comprised of a number of underlying feeds that use say 6 decimals but the resultant feed is set to 18 decimals. Or, if running on an old emp, it is possible the decimals don't match underlying feeds.

In ether case, do you think it worthwhile to add a constructor param that scales the output according to some `priceFeedDecimals` in the same way we treat other feed types?

mrice32 replied:
It's a fair point. Will add.

chrismaree reviewed (approved):
I had a few nits/comments and questions but overall this is epic and super powerful! Great work.

mrice32 commented on packages/financial-templates-lib/src/price-feed/ExpressionPriceFeed.js:
```diff
@@ -0,0 +1,117 @@
+  precision: 100
+});
+
+// Gets balancer spot and historical prices. This price feed assumes that it is returning
```
Ha, yeah, this is just copy paste :) will fix

pemulis commented on packages/financial-templates-lib/package.json:
```diff
@@ -11,6 +11,7 @@
     "@uniswap/sdk": "^2.0.5",
     "bluebird": "^3.7.2",
     "dotenv": "^6.2.0",
+    "mathjs": "^9.2.0",
```
As a general comment about smart contract interface development patterns, and not this pull request specifically, I wonder if it would be better for interfaces that are mostly reading and writing data from smart contracts to do most of their calculation (and maybe rendering) on the server side.

This came up in the engineering team discussion, where we determined that asking people who use this package to run a server would be a heavy lift, versus doing calculations in the browser (which `mathjs` allows for) or hitting an API we provide, which would be a centralized point of failure in addition to taking a lot of time to build and maintain.

In the long run, I think that smart contract front-ends that pull most of the work out of the client and put it on the server will win out, since very little of what you want to do with smart contracts resembles the social media clients that React _et al._ were really designed for. Every user interaction takes at least a few seconds to finalize, and data updates also take a few seconds. Doing most of the stuff on the server side with a light client-side interface feels like the right trade-off.

I suppose this is also some background for the general community on why we're going in this direction, since `mathjs`, while quite heavy, will work in the browser, while some other alternatives we considered would not.

mrice32 replied:
Yeah, I could imagine employing a much heavier server-side implementation. If we really wanted our UIs to be resilient, they could have a fallback setup where they _only_ pull down the price feed js code if the backend/server-side is down for some reason. This, while complicated, could be the best of both worlds, although I doubt we'll ever have the engineering resources to build something like that :)

pemulis commented on packages/financial-templates-lib/src/price-feed/ExpressionPriceFeed.js:
```diff
@@ -0,0 +1,114 @@
+// Gets balancer spot and historical prices. This price feed assumes that it is returning
+// prices as 18 decimals of precision, so it will scale up the pool's price as reported by Balancer contracts
+// if the user specifies that the Balancer contract is returning non-18 decimal precision prices.
+class ExpressionPriceFeed extends PriceFeedInterface {
```
Another meta comment: If you wanted a price feed that is the result of two existing, approved price feeds run through this calculator, does that price feed need to be approved through the UMIP process, or would people be free to do price feed mash-ups in new synthetics without an extra vote?

mrice32 replied:
@pemulis this is a great question. To put it simply, the latter. However, I want to make a larger point. None of our price feeds know anything about approved identifiers. They only know about the default configs in the price feed directory. Those default configs allow us to automatically construct price feeds based on the identifier the EMP is using without needing a custom config input. When an UMIP is approved to add an identifier, we generally add a default config for it.

In addition to expressions automatically being able to use default configs with no additional information, you will also be able to specify a `customFeeds` map inside the ExpressionPriceFeed config that maps your own custom symbols so you're not limited to the default configs. Here's an example of a working config:
```js
{
  type: "expression",
  expression: "ETHBTC + MYCUSTOMFEED",
  customFeeds: {
    MYCUSTOMFEED: {
      type: "cryptowatch",
      exchange: "coinbase-pro",
      pair: "ethusd"
    }
  }
}
```

pemulis replied:
The only thing actually on-chain is the identifier string, right?

mrice32 replied:
Yep!

mrice32 commented:
@chrismaree @nicholaspai I re-requested y'all's reviews. I added the configuration/construction code, which is pretty important for this sort of price feed that's primarily meant for making configuration and construction easier. I added this to the description, but copying below so you see it.

This is how you configure an expression price feed:
```js
{
  type: "expression",
  expression: "ETHBTC + MYCUSTOMFEED",
  customFeeds: {
    MYCUSTOMFEED: {
      type: "cryptowatch",
      exchange: "coinbase-pro",
      pair: "ethusd"
    }
  }
}
```

In this example, the parser would find that the two "symbols" in the expression are `ETHBTC` (notice that we have stripped the `/` character) and `MYCUSTOMFEED`. Preferring customFeeds in the resolution, we would look through customFeeds _and_ the default feeds in the repo searching for each of these symbols. If any resolution fails, the construction fails. In this case, however, it would construct both feeds and use them to evaluate the expression. The nice thing about the early resolution is that any symbols that aren't in the expression are pruned, so there should never be any unnecessary feeds being constructed and/or evaluated.

chrismaree commented:
> @chrismaree @nicholaspai I re-requested y'all's reviews. I added the configuration/construction code, which is pretty important for this sort of price feed that's primarily meant for making configuration and construction easier. I added this to the description, but copying below so you see it.
> 
> This is how you configure an expression price feed:
> 
> ```js
> {
>   type: "expression",
>   expression: "ETHBTC + MYCUSTOMFEED",
>   customFeeds: {
>     MYCUSTOMFEED: {
>       type: "cryptowatch",
>       exchange: "coinbase-pro",
>       pair: "ethusd"
>     }
>   }
> }
> ```
> 
> In this example, the parser would find that the two "symbols" in the expression are `ETHBTC` (notice that we have stripped the `/` character, which is a requirement because the parser will interpret it as division) and `MYCUSTOMFEED`. Preferring customFeeds in the resolution, we would look through customFeeds _and_ the default feeds in the repo searching for each of these symbols. If any resolution fails, the construction fails. In this case, however, it would construct both feeds and use them to evaluate the expression. The nice thing about the early resolution is that any symbols that aren't in the expression are pruned, so there should never be any unnecessary feeds being constructed and/or evaluated.

This is so cool. only real downside I can see is stripping out the `/` char as now your feed won't match the actual price identifiers.

The ability to define a custom feed like this is epic. This means you can define as many constituent feeds that you need then define the `expression` to join them together.

chrismaree commented on packages/financial-templates-lib/src/price-feed/CreatePriceFeed.js:
```diff
@@ -174,6 +189,65 @@ async function createPriceFeed(logger, web3, networker, getTime, config) {
+    // customFeeds with the same name. Tranform keys by stripping "/" since that would be interpreted as division.
+    const allConfigs = Object.fromEntries(
+      Object.entries({ ...defaultConfigs, ...expressionConfig.customFeeds }).map(([key, value]) => {
+        return [key.replace("/", ""), value];
```
hold on. just to double-check I understand. I can't put any feed names that have `/` within the feed expression, right? this is just parsing the list of default feeds to match this rule that expression feeds cant contain any `/`?

mrice32 replied:
The way it previously worked (and there was a subtle bug :)) was that you would enter an expression, like:
```
"ETHBTC + STABLESPREADUSDC"
```

That would match the two existing identifiers:
```
ETH/BTC
STABLESPREAD/USDC
```
because we strip out the `/` before searching through the default list.

The weird caveat is that it would also match:
```
ETHBTC
ET/HBTC
E/THBTC
...
```

This has now changed to using escape characters. PTAL :)

chrismaree commented on packages/financial-templates-lib/src/price-feed/CreatePriceFeed.js:
```diff
@@ -174,6 +189,65 @@ async function createPriceFeed(logger, web3, networker, getTime, config) {
+      })
+    );
+
+    // This call chain:
```
is there any way to detect if the user has incorrectly placed a `/` within the feed within an expression? I guess if it parses this as deciding two parts of a feed it would error out? ETH/BTC would be ETH divided by BTC and we don't have a feed that is just `ETH` or `BTC` so it would error, right?

mrice32 replied:
Exactly. 💯 . That's effectively our only way to verify the correctness -- rely on mathjs to tell us if there are any unresolved symbols.

daywiss commented on packages/financial-templates-lib/src/price-feed/CreatePriceFeed.js:
```diff
@@ -174,6 +188,66 @@ async function createPriceFeed(logger, web3, networker, getTime, config) {
+
+          // If there is no config for this symbol, insert null and send an error.
+          if (!config) {
+            logger.error({
```
why doesnt this just throw an error, or log error and throw error?

mrice32 replied:
This is how the others work I believe -- just continuing the pattern from above where we log an error and return null.

daywiss commented on packages/financial-templates-lib/src/price-feed/CreatePriceFeed.js:
```diff
@@ -174,6 +188,66 @@ async function createPriceFeed(logger, web3, networker, getTime, config) {
+    );
+
+    // Return null if any of the price feeds in the map are null (meaning there was an error).
+    if (Object.values(priceFeedMap).some(priceFeed => priceFeed === null)) return null;
```
if error was thrown earlier, i dont think this would be needed

mrice32 replied:
Agreed. However, the current pattern is to return null and log an error. I agree that this may be a bit nonsensical looking back, but I would prefer to update this pattern in a follow-up PR across this entire file rather than 1) being inconsistent or 2) creating a lot of unrelated churn by changing all of the creation functions.

daywiss commented on packages/financial-templates-lib/src/price-feed/ExpressionPriceFeed.js:
```diff
@@ -0,0 +1,154 @@
+    const errors = [];
+    await Promise.all(
+      Object.entries(this.priceFeedMap).map(async ([name, pf]) => {
+        const price = await pf.getHistoricalPrice(time).catch(err => errors.push(err));
```
why accumulate errors here rather than throwing on the first one? you cant continue if you have a single error anyway

mrice32 replied:
This is the way we do this in the medianizer feed, so I just continued the pattern. The idea is that _all_ lower-level feed errors should be bubbled up for the purpose of logging in case multiple are failing. This has been useful for the complex stablespread feeds.

daywiss commented on packages/financial-templates-lib/src/price-feed/ExpressionPriceFeed.js:
```diff
@@ -0,0 +1,154 @@
+      }
+    })
+    .join("");
+}
```
this function looks tricky, you might want a test for this, though im pretty sure its working by proxy of testing everything else

mrice32 replied:
It has a test (not sure if this was true when you reviewed).

mrice32 replied:
Sorry, I just now realized that I forgot to push a few changes, but there's definitely a test in there now :)

daywiss commented on packages/financial-templates-lib/test/price-feed/CreatePriceFeed.js:
```diff
@@ -713,6 +713,119 @@ contract("CreatePriceFeed.js", function(accounts) {
+
+    const expressionPriceFeed = await createPriceFeed(logger, web3, networker, getTime, config);
+
+    assert.isNotNull(expressionPriceFeed);
```
nice test suite, wondering if its worth also testing for the error log?

mrice32 replied:
Done!

daywiss commented on packages/financial-templates-lib/src/price-feed/ExpressionPriceFeed.js:
```diff
@@ -0,0 +1,154 @@
+
+    if (errors.length > 0) return null;
+
+    return this._convertToFixed(this.expressionCode.evaluate(prices));
```
wonder if we can make this DRYer since the logic is really similar to getting historical price. its almost like you could create a higher order function which takes in a `getPrice` call which can represent getCurrentPrice or getHistorical price. wdyt?

mrice32 replied:
Two issues I see with that approach (the first likely making this a nonstarter):
- getHistoricalPrice is async and this higher level function would likely need to be async as a result of calling getHistoricalPrice on other feeds. If this function is async, getCurrentPrice can't call it. _Technically_ there may be ways around this by trying to do the async stuff outside of the function, but I think that might be so complex as to defeat the purpose of this refactor.
- I suspect this may hurt readability. I'm not sure, though (I'd have to try to implement it to really tell).

Thoughts?

daywiss replied:
yes the sync/async issue does break that idea! thanks for thinking about this

daywiss commented on packages/financial-templates-lib/test/price-feed/ExpressionPriceFeed.js:
```diff
@@ -0,0 +1,163 @@
+
+    // Should return the sum of the current prices.
+    // (1 + 9) * (2 + 9) / 7.321 = 15.025269771889086190 (rounded to 18 decimals).
+    assert.equal(expressionPriceFeed.getCurrentPrice(), toWei("15.025269771889086190"));
```
really cool to see this in action!

daywiss commented on packages/financial-templates-lib/test/price-feed/ExpressionPriceFeed.js:
```diff
@@ -0,0 +1,163 @@
+      const escapedSymbol = "\\[1\\/\\-2\\]\\ \\-\\ 17";
+      assert.equal(math.evaluate(`${escapedSymbol} - 2 * ${escapedSymbol}`, { [escapedSymbol]: 5 }).toString(), "-5");
+
+      const unescapedSymbol = "USD-[bwBTC/ETH SLP]";
```
would have been good for the tests above to use this identifier in one of the expressions

mrice32 replied:
Done! Added it to one of the earlier tests.

daywiss reviewed (commented):
really interesting approach and solution. you've touched on the implications of using this pattern which could be quite powerful!

chrismaree commented on packages/financial-templates-lib/test/price-feed/ExpressionPriceFeed.js:
```diff
@@ -0,0 +1,163 @@
+
+  describe("Expression parsing with escaped characters", async function() {
+    it("Escapes characters correctly", async function() {
+      assert.equal(escapeSpecialCharacters("USD-[bwBTC/ETH SLP]"), "USD\\-\\[bwBTC\\/ETH\\ SLP\\]");
```
nice.

chrismaree reviewed (approved):
overall this looks great to me! only issue I can see is that currently you are not implementing `getHistoricalPricePeriods` so this won't work out the box with developer mining. However, this is probs fine as for now we dont have all feeds implementing this method always.

mrice32 commented on packages/financial-templates-lib/src/price-feed/CreatePriceFeed.js:
```diff
@@ -174,6 +188,66 @@ async function createPriceFeed(logger, web3, networker, getTime, config) {
+    // Build list of configs that could be used in the expression including default price feed configs and customFeeds
+    // that the user has provided inside the ExpressionPriceFeed config. Note: default configs are overriden by
+    // customFeeds with the same name. Tranform keys by escaping any special characters in the identifier names..
+    const allConfigs = Object.fromEntries(
```
I hadn't either! It's a pretty new function from what I can tell. I doubt we could use it on node12 :)

mrice32 commented:
> only issue I can see is that currently you are not implementing `getHistoricalPricePeriods` so this won't work out the box with developer mining. However, this is probs fine as for now we dont have all feeds implementing this method always.

@chrismaree it's a great point. I think there are two good reasons for _not_ doing this:
- It's not clear that we're actually going to keep that method medium-term. It seems like we may be moving towards historical capture for dev mining.
- Not all constituent price feeds implement it, so we would have to assume that the passed in price feeds implement to call them even though they might not. I guess the consequence is that we'd have a method that will _sometimes_ always error depending on the sub-price feeds that you use in your expression.

daywiss reviewed (approved):
nice work, thanks for updates
